File: app/services/similarity_service.py
"""Service for detecting duplicate and similar questions using embeddings."""

import hashlib
from typing import List, Tuple, Optional
import numpy as np
from sqlalchemy.orm import Session
from app.models import Question
import os
from dotenv import load_dotenv

load_dotenv()

# Load similarity threshold from environment
SIMILARITY_THRESHOLD = float(os.getenv("SIMILARITY_THRESHOLD", "0.85"))

# Initialize sentence transformer model (lazy loading)
_model = None


def get_model():
    """
    Get or initialize the sentence transformer model.
    
    Lazy import to avoid import errors at module level.

    Returns:
        SentenceTransformer: The model instance.
        
    Raises:
        ImportError: If sentence_transformers is not properly installed.
    """
    global _model
    if _model is None:
        try:
            from sentence_transformers import SentenceTransformer
            # Using a lightweight model suitable for POC
            _model = SentenceTransformer("all-MiniLM-L6-v2")
        except ImportError as e:
            raise ImportError(
                "sentence-transformers is not properly installed. "
                "Please run: pip install --upgrade sentence-transformers huggingface-hub"
            ) from e
    return _model


def normalize_text(text: str) -> str:
    """
    Normalize text for hashing (lowercase, strip whitespace).

    Args:
        text (str): Input text.

    Returns:
        str: Normalized text.
    """
    return text.lower().strip()


def generate_hash(question_text: str) -> str:
    """
    Generate SHA256 hash for exact duplicate detection.

    Args:
        question_text (str): Question text.

    Returns:
        str: SHA256 hash of normalized question text.
    """
    normalized = normalize_text(question_text)
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()


def generate_embedding(question_text: str) -> List[float]:
    """
    Generate embedding vector for similarity detection.

    Args:
        question_text (str): Question text.

    Returns:
        List[float]: Embedding vector as a list.
    """
    model = get_model()
    embedding = model.encode(question_text, convert_to_numpy=True)
    return embedding.tolist()


def cosine_similarity(vec1: List[float], vec2: List[float]) -> float:
    """
    Calculate cosine similarity between two vectors.

    Args:
        vec1 (List[float]): First embedding vector.
        vec2 (List[float]): Second embedding vector.

    Returns:
        float: Cosine similarity score between 0 and 1.
    """
    v1 = np.array(vec1)
    v2 = np.array(vec2)
    dot_product = np.dot(v1, v2)
    norm1 = np.linalg.norm(v1)
    norm2 = np.linalg.norm(v2)
    if norm1 == 0 or norm2 == 0:
        return 0.0
    return float(dot_product / (norm1 * norm2))


def check_exact_duplicate(db: Session, question_hash: str, college: Optional[str] = None) -> Optional[Question]:
    """
    Check if a question with the same hash already exists.
    If college is provided, only checks within that college.

    Args:
        db (Session): Database session.
        question_hash (str): Hash of the question text.
        college (Optional[str]): College name to filter by. If provided, only checks duplicates in this college.

    Returns:
        Optional[Question]: Existing question if found, None otherwise.
    """
    query = db.query(Question).filter(
        Question.question_hash == question_hash,
        Question.status == "Active"
    )
    if college:
        query = query.filter(Question.college == college)
    return query.first()


def find_similar_questions(
    db: Session, embedding: List[float], threshold: float = SIMILARITY_THRESHOLD, exclude_id: Optional[int] = None, college: Optional[str] = None
) -> List[Tuple[Question, float]]:
    """
    Find questions with similar embeddings above the threshold.
    If college is provided, only checks within that college.

    Args:
        db (Session): Database session.
        embedding (List[float]): Embedding vector to compare against.
        threshold (float): Similarity threshold (default from env).
        exclude_id (Optional[int]): Question ID to exclude from results.
        college (Optional[str]): College name to filter by. If provided, only checks duplicates in this college.

    Returns:
        List[Tuple[Question, float]]: List of (question, similarity_score) tuples.
    """
    # Get all active questions with embeddings
    query = db.query(Question).filter(
        Question.status == "Active",
        Question.embedding.isnot(None)
    )
    
    if exclude_id:
        query = query.filter(Question.id != exclude_id)
    
    if college:
        query = query.filter(Question.college == college)
    
    questions = query.all()
    
    similar_questions = []
    for question in questions:
        if question.embedding:
            similarity = cosine_similarity(embedding, question.embedding)
            if similarity >= threshold:
                similar_questions.append((question, similarity))
    
    # Sort by similarity score (descending)
    similar_questions.sort(key=lambda x: x[1], reverse=True)
    
    return similar_questions


def check_similarity(
    db: Session, question_text: str, exclude_id: Optional[int] = None, college: Optional[str] = None
) -> Tuple[bool, bool, List[Tuple[Question, float]]]:
    """
    Check if a question is duplicate or similar to existing questions.
    If college is provided, only checks within that college.

    Args:
        db (Session): Database session.
        question_text (str): Question text to check.
        exclude_id (Optional[int]): Question ID to exclude from results.
        college (Optional[str]): College name to filter by. If provided, only checks duplicates in this college.

    Returns:
        Tuple[bool, bool, List[Tuple[Question, float]]]:
            - is_duplicate: True if exact or similar match found
            - exact_match: True if exact match found
            - similar_questions: List of (question, similarity_score) tuples
    """
    # Check for exact duplicate
    question_hash = generate_hash(question_text)
    exact_match_question = check_exact_duplicate(db, question_hash, college=college)
    
    exact_match = exact_match_question is not None
    
    # Generate embedding for similarity check
    embedding = generate_embedding(question_text)
    
    # Find similar questions
    similar_questions = find_similar_questions(db, embedding, exclude_id=exclude_id, college=college)
    
    # Consider it a duplicate if exact match or similar questions found
    is_duplicate = exact_match or len(similar_questions) > 0
    
    # Include exact match in similar questions if found
    if exact_match and exact_match_question:
        # Check if exact match is already in similar_questions
        if not any(q.id == exact_match_question.id for q, _ in similar_questions):
            similar_questions.insert(0, (exact_match_question, 1.0))
    
    return is_duplicate, exact_match, similar_questions

File: backend/app/services/similarityService.js
/**
 * Service for detecting duplicate and similar questions using embeddings.
 */

import crypto from "crypto";
import { pipeline } from "@xenova/transformers";

// Load similarity threshold from environment
const SIMILARITY_THRESHOLD = parseFloat(process.env.SIMILARITY_THRESHOLD || "0.85");

// Lazy load the embedding model
let embeddingModel = null;

/**
 * Get or initialize the embedding model.
 * Uses @xenova/transformers for browser/Node.js compatibility.
 *
 * @returns {Promise<Object>} The embedding model pipeline
 */
let modelLoadingPromise = null;

async function getModel() {
  if (!embeddingModel) {
    if (!modelLoadingPromise) {
      modelLoadingPromise = (async () => {
        try {
          // Use the same model as Python version: all-MiniLM-L6-v2
          console.log("Loading embedding model: all-MiniLM-L6-v2 (this may take a moment on first run)...");
          embeddingModel = await pipeline(
            "feature-extraction",
            "Xenova/all-MiniLM-L6-v2"
          );
          console.log("Embedding model loaded successfully");
          return embeddingModel;
        } catch (error) {
          modelLoadingPromise = null; // Reset on error
          throw new Error(
            `Failed to load embedding model: ${error.message}. ` +
              "Please ensure @xenova/transformers is properly installed."
          );
        }
      })();
    }
    await modelLoadingPromise;
  }
  return embeddingModel;
}

/**
 * Normalize text for hashing (lowercase, strip whitespace).
 *
 * @param {string} text - Input text
 * @returns {string} Normalized text
 */
export function normalizeText(text) {
  return text.toLowerCase().trim();
}

/**
 * Generate SHA256 hash for exact duplicate detection.
 *
 * @param {string} questionText - Question text
 * @returns {string} SHA256 hash of normalized question text
 */
export function generateHash(questionText) {
  const normalized = normalizeText(questionText);
  return crypto.createHash("sha256").update(normalized, "utf-8").digest("hex");
}

/**
 * Generate embedding vector for similarity detection.
 *
 * @param {string} questionText - Question text
 * @returns {Promise<number[]>} Embedding vector as an array
 */
export async function generateEmbedding(questionText) {
  const model = await getModel();
  const output = await model(questionText, { pooling: "mean", normalize: true });
  // Convert tensor to array
  return Array.from(output.data);
}

/**
 * Calculate cosine similarity between two vectors.
 *
 * @param {number[]} vec1 - First embedding vector
 * @param {number[]} vec2 - Second embedding vector
 * @returns {number} Cosine similarity score between 0 and 1
 */
export function cosineSimilarity(vec1, vec2) {
  if (vec1.length !== vec2.length) {
    return 0.0;
  }

  let dotProduct = 0;
  let norm1 = 0;
  let norm2 = 0;

  for (let i = 0; i < vec1.length; i++) {
    dotProduct += vec1[i] * vec2[i];
    norm1 += vec1[i] * vec1[i];
    norm2 += vec2[i] * vec2[i];
  }

  norm1 = Math.sqrt(norm1);
  norm2 = Math.sqrt(norm2);

  if (norm1 === 0 || norm2 === 0) {
    return 0.0;
  }

  return dotProduct / (norm1 * norm2);
}

/**
 * Check if a question with the same hash already exists.
 *
 * @param {import('@supabase/supabase-js').SupabaseClient} supabase - Supabase client
 * @param {string} questionHash - Hash of the question text
 * @param {string|null} college - Optional college name to filter by
 * @returns {Promise<Object|null>} Existing question if found, null otherwise
 */
export async function checkExactDuplicate(supabase, questionHash, college = null) {
  let query = supabase
    .from("questions")
    .select("*")
    .eq("question_hash", questionHash)
    .eq("status", "Active")
    .limit(1);

  if (college) {
    query = query.eq("college", college);
  }

  const { data, error } = await query.single();

  if (error && error.code !== "PGRST116") {
    // PGRST116 is "not found" - that's okay
    throw error;
  }

  return data || null;
}

/**
 * Find questions with similar embeddings above the threshold.
 *
 * @param {import('@supabase/supabase-js').SupabaseClient} supabase - Supabase client
 * @param {number[]} embedding - Embedding vector to compare against
 * @param {number} threshold - Similarity threshold (default from env)
 * @param {number|null} excludeId - Optional question ID to exclude
 * @param {string|null} college - Optional college name to filter by
 * @returns {Promise<Array<{question: Object, score: number}>>} List of similar questions with scores
 */
export async function findSimilarQuestions(
  supabase,
  embedding,
  threshold = SIMILARITY_THRESHOLD,
  excludeId = null,
  college = null
) {
  // Get all active questions with embeddings
  let query = supabase
    .from("questions")
    .select("*")
    .eq("status", "Active")
    .not("embedding", "is", null);

  if (excludeId) {
    query = query.neq("id", excludeId);
  }

  if (college) {
    query = query.eq("college", college);
  }

  const { data: questions, error } = await query;

  if (error) {
    throw error;
  }

  const similarQuestions = [];

  for (const question of questions || []) {
    if (question.embedding && Array.isArray(question.embedding)) {
      const similarity = cosineSimilarity(embedding, question.embedding);
      if (similarity >= threshold) {
        similarQuestions.push({ question, score: similarity });
      }
    }
  }

  // Sort by similarity score (descending)
  similarQuestions.sort((a, b) => b.score - a.score);

  return similarQuestions;
}

/**
 * Check if a question is duplicate or similar to existing questions.
 *
 * @param {import('@supabase/supabase-js').SupabaseClient} supabase - Supabase client
 * @param {string} questionText - Question text to check
 * @param {number|null} excludeId - Optional question ID to exclude
 * @param {string|null} college - Optional college name to filter by
 * @returns {Promise<{isDuplicate: boolean, exactMatch: boolean, similarQuestions: Array}>}
 */
export async function checkSimilarity(
  supabase,
  questionText,
  excludeId = null,
  college = null
) {
  // Check for exact duplicate
  const questionHash = generateHash(questionText);
  const exactMatchQuestion = await checkExactDuplicate(supabase, questionHash, college);

  const exactMatch = exactMatchQuestion !== null;

  // Generate embedding for similarity check
  const embedding = await generateEmbedding(questionText);

  // Find similar questions
  const similarQuestions = await findSimilarQuestions(
    supabase,
    embedding,
    SIMILARITY_THRESHOLD,
    excludeId,
    college
  );

  // Consider it a duplicate if exact match or similar questions found
  const isDuplicate = exactMatch || similarQuestions.length > 0;

  // Include exact match in similar questions if found
  if (exactMatch && exactMatchQuestion) {
    // Check if exact match is already in similarQuestions
    const alreadyIncluded = similarQuestions.some(
      (item) => item.question.id === exactMatchQuestion.id
    );
    if (!alreadyIncluded) {
      similarQuestions.unshift({ question: exactMatchQuestion, score: 1.0 });
    }
  }

  return {
    isDuplicate,
    exactMatch,
    similarQuestions: similarQuestions.map((item) => ({
      question: item.question,
      score: item.score,
    })),
  };
}

